BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java

```diff
@@ -116,6 +116,47 @@ public T get(BlueKey key) throws BlueDbException {
 		return results;
 	}
 
+	// TODO make private?  or somehow enforce standard levels
+	public void rollup(long start, long end) throws BlueDbException {
+		List<File> filesToRollup = getFilesInRange(start, end);
+		Path path = Paths.get(segmentPath.toString(), start + "_" + end);
+		Path tmpPath = FileManager.createTempFilePath(path);
+		List<BlueEntity<T>> entities = new ArrayList<>();
+
+		// TODO switch to streaming
+		// TODO keep sorted, here and everywhere you write
+		for (File file: filesToRollup) {
+			List<BlueEntity<T>> es = fetch(file);
+			entities.addAll(es);
+		}
+		LockManager<Path> lockManager = fileManager.getLockManager();
+		try (BlueWriteLock<Path> tempFileLock = lockManager.acquireWriteLock(tmpPath)) {
+			fileManager.saveList(tempFileLock, entities);
+		}
+
+		List<BlueWriteLock<Path>> sourceFileWriteLocks = new ArrayList<>();
+		try (BlueWriteLock<Path> targetFileLock = lockManager.acquireWriteLock(path)){
+			for (File file: filesToRollup) {
+				sourceFileWriteLocks.add(lockManager.acquireWriteLock(file.toPath()));
+			}
+
+			FileManager.moveFile(tmpPath, targetFileLock);
+			for (BlueWriteLock<Path> writeLock: sourceFileWriteLocks) {
+				FileManager.deleteFile(writeLock);

```

Comment:
Could something be added to one of the files being rolled up after we've already added all of its entities to our list? If so, they didn't make it into the temp file and we'll lose them.

Reply:
not right now.  the rollup is currently done in the write thread.  If we change that, we'll have to change a bunch of logic.